### PR TITLE
Fix statefulMap may call onComplete twice if it throws

### PR DESCRIPTION
**Pull request by GreyPlane**
#596

### CODE REVIEW

File: stream/src/main/scala/org/apache/pekko/stream/impl/fusing/Ops.scala [hunk 1]
```diff
@@ -2273,25 +2273,27 @@ private[pekko] final class StatefulMap[S, In, Out](create: () => S, f: (S, In) =
...
           case None       => failStage(ex)
         }
-        needInvokeOnCompleteCallback = false
       }
 
       override def onPull(): Unit = pull(in)
 
       override def postStop(): Unit = {
         if (needInvokeOnCompleteCallback) {
+          println("changed 2")
```
Comment (pjfanning) [inline stream/src/main/scala/org/apache/pekko/stream/impl/fusing/Ops.scala:2296]: can you remove the printlns?

**Review by pjfanning — CHANGES_REQUESTED** [review]

File: stream-tests/src/test/scala/org/apache/pekko/stream/scaladsl/FlowStatefulMapSpec.scala [hunk 2]
```diff
@@ -35,6 +36,8 @@ class FlowStatefulMapSpec extends StreamSpec {
 
   val ex = new Exception("TEST") with NoStackTrace
 
+  val dtorException = new Exception("bad dtor") with NoStackTrace
```
Comment (He-Pin) [inline stream-tests/src/test/scala/org/apache/pekko/stream/scaladsl/FlowStatefulMapSpec.scala:39]: dtor means `destructor`?

**Comment by He-Pin** [comment]
LGTM only two comments.

File: stream-tests/src/test/scala/org/apache/pekko/stream/scaladsl/FlowStatefulMapSpec.scala [hunk 3]
```diff
@@ -280,4 +283,64 @@ class FlowStatefulMapSpec extends StreamSpec {
...
+          badDtor()
+          None
+        })
+      .runWith(Sink.ignore)
+
+    StreamTestKit.assertAllStagesStopped(None)
+    counter shouldBe 1
+  }
+
+  "won't call onComplete twice when upstream finished even if onComplete throws" in {
```
Comment (He-Pin) [inline stream-tests/src/test/scala/org/apache/pekko/stream/scaladsl/FlowStatefulMapSpec.scala:327]: `in assertAllStagesStopped {`

**Comment by pjfanning** [comment]
@GreyPlane could we get you to fill in an Apache CLA for this? https://www.apache.org/licenses/contributor-agreements.html

For a comparatively small change like this, we might not always require the CLA but this issue/PR has become a potential legal issue since @kerr has looked at the Akka solution (and it is not licensed under a license that is compatible with use by us). We don't want any risk that anyone could claim that Category X code has been copied into our project.

https://www.apache.org/legal/resolved.html#category-x

**Comment by He-Pin** [comment]
@pjfanning The fix of my came out independently,  but the tests were what I copied for quick validation, let's make that clear, and then i checked the Akka fix which is not the same as mine.
So it's the tests I took not the fix.

Many ones in Wechat group can prove this.

I was want to make sure the onComplete to be called and then mark the boolean to false, but that will need the onComplete be idempotent and not easy to use.

**Comment by pjfanning** [comment]
@kerr Let us do a clean room fix.

**Comment by He-Pin** [comment]
And you may find someone not in tech to check and validation those fix is not the same

**Comment by He-Pin** [comment]
@GreyPlane Sorry, I think this fix should be closed, even I knew you come up this independently, but as you are involved/leved comments in the akka pr, so this fix should be closed and come out with someone else, thank you very much.

**Pull request closed without merging**